File: backend/app/schemas/chain_of_custody.py
from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, Field

from app.schemas.integrity import IntegrityMismatch


IntegrityState = Literal[
    "VERIFIED",
    "MISSING_ON_CHAIN",
    "ORPHANED_ON_CHAIN",
    "INTEGRITY_MISMATCH",
    "LEGACY_PARTIAL_VERIFICATION",
    "BLOCKCHAIN_UNAVAILABLE",
]


class ChainUserIdentity(BaseModel):
    user_id: UUID
    display_name: str
    role: str
    badge_number: str | None = None
    username: str
    email: str
    full_name: str | None = None
    rank: str | None = None


class ChainTransactionMetadata(BaseModel):
    tx_hash: str | None
    block_number: int | None
    status: str
    verified: bool


class ChainEvidenceMetadata(BaseModel):
    evidence_id: UUID
    evidence_number: str
    evidence_ref: str
    evidence_hash: str | None
    original_sha256: str | None
    uploaded_at: datetime | None
    blockchain_recorded_at: int | None
    writer: str | None
    registration_tx_hash: str | None = None
    registration_block_number: int | None = None
    registration_transaction_index: int | None = None
    registration_log_index: int | None = None


class ChainAccessMetadata(BaseModel):
    evidence_ref: str
    officer_ref: str
    access_session_ref: str
    action: str
    occurred_at: int
    recorded_at: int
    writer: str
    transaction_hash: str
    block_number: int
    transaction_index: int
    log_index: int


class ChainAccessDatabaseMetadata(BaseModel):
    access_log_id: UUID
    evidence_id: UUID | None
    user_id: UUID
    action: str
    accessed_at: datetime
    tx_internal_id: UUID | None


class ChainIntegrityMismatch(IntegrityMismatch):
    explanation: str


class ChainAccessVerification(BaseModel):
    session_exists: bool
    access_log_exists: bool
    transaction_exists: bool
    evidence_ref_matches: bool
    officer_ref_matches: bool
    action_matches: bool
    occurred_at_matches: bool
    transaction_matches: bool


class ChainAccessHistoryItem(BaseModel):
    access_log_id: UUID | None
    access_session_ref: str
    user: ChainUserIdentity | None
    database_user: ChainUserIdentity | None
    action: str
    accessed_at: datetime | None
    database: ChainAccessDatabaseMetadata | None
    blockchain: ChainAccessMetadata | None
    transaction: ChainTransactionMetadata | None
    verified: bool
    integrity_state: IntegrityState
    verification: ChainAccessVerification
    mismatches: list[ChainIntegrityMismatch] = Field(default_factory=list)


class ChainOfCustodyVerification(BaseModel):
    evidence_exists: bool
    evidence_hash_matches: bool
    uploader_ref_matches: bool
    registration_transaction_matches: bool
    access_records_verified: int
    access_records_total: int


class ChainOfCustodyResponse(BaseModel):
    verified: bool
    integrity_state: IntegrityState
    evidence: ChainEvidenceMetadata
    uploader: ChainUserIdentity | None
    registration_transaction: ChainTransactionMetadata | None
    access_history: list[ChainAccessHistoryItem]
    # หลักฐานที่ถูกเข้าถึงบ่อยมีประวัติได้หลายพันรายการ — access_history เป็นเพียงหน้าเดียว
    # ส่วน verification/verified ยังคำนวณจากประวัติทั้งหมดเสมอ
    access_history_total: int = 0
    access_history_limit: int | None = None
    access_history_offset: int = 0
    verification: ChainOfCustodyVerification

File: frontend/src/interfaces/chainOfCustody.ts
export interface ChainUserIdentity {
  user_id: string;
  display_name: string;
  role: string;
  badge_number: string | null;
  username: string;
  email: string;
  full_name: string | null;
  rank: string | null;
}

export interface ChainTransactionMetadata {
  tx_hash: string | null;
  block_number: number | null;
  status: string;
  verified: boolean;
}

export interface ChainEvidenceMetadata {
  evidence_id: string;
  evidence_number: string;
  evidence_ref: string;
  evidence_hash: string | null;
  original_sha256: string | null;
  uploaded_at: string | null;
  blockchain_recorded_at: number | null;
  writer: string | null;
  registration_tx_hash: string | null;
  registration_block_number: number | null;
  registration_transaction_index: number | null;
  registration_log_index: number | null;
}

export interface ChainAccessMetadata {
  evidence_ref: string;
  officer_ref: string;
  access_session_ref: string;
  action: string;
  occurred_at: number;
  recorded_at: number;
  writer: string;
  transaction_hash: string;
  block_number: number;
  transaction_index: number;
  log_index: number;
}

export interface ChainAccessDatabaseMetadata {
  access_log_id: string;
  evidence_id: string | null;
  user_id: string;
  action: string;
  accessed_at: string;
  tx_internal_id: string | null;
}

export interface ChainAccessVerification {
  session_exists: boolean;
  access_log_exists: boolean;
  transaction_exists: boolean;
  evidence_ref_matches: boolean;
  officer_ref_matches: boolean;
  action_matches: boolean;
  occurred_at_matches: boolean;
  transaction_matches: boolean;
}

export interface IntegrityMismatch {
  field: string;
  database_value: unknown;
  blockchain_value: unknown;
  explanation?: string;
}

export type ChainIntegrityState =
  | "VERIFIED"
  | "MISSING_ON_CHAIN"
  | "ORPHANED_ON_CHAIN"
  | "INTEGRITY_MISMATCH"
  | "LEGACY_PARTIAL_VERIFICATION"
  | "BLOCKCHAIN_UNAVAILABLE";

export interface ChainAccessHistoryItem {
  access_log_id: string | null;
  access_session_ref: string;
  user: ChainUserIdentity | null;
  database_user: ChainUserIdentity | null;
  action: string;
  accessed_at: string | null;
  database: ChainAccessDatabaseMetadata | null;
  blockchain: ChainAccessMetadata | null;
  transaction: ChainTransactionMetadata | null;
  verified: boolean;
  integrity_state: ChainIntegrityState;
  verification: ChainAccessVerification;
  mismatches: IntegrityMismatch[];
}

export interface ChainOfCustodyVerification {
  evidence_exists: boolean;
  evidence_hash_matches: boolean;
  uploader_ref_matches: boolean;
  registration_transaction_matches: boolean;
  access_records_verified: number;
  access_records_total: number;
}

export interface ChainOfCustodyResponse {
  verified: boolean;
  integrity_state: ChainIntegrityState;
  evidence: ChainEvidenceMetadata;
  uploader: ChainUserIdentity | null;
  registration_transaction: ChainTransactionMetadata | null;
  access_history: ChainAccessHistoryItem[];
  /** จำนวนรายการทั้งหมด — access_history เป็นเพียงหน้าเดียว ส่วน verification คิดจากทั้งหมด */
  access_history_total: number;
  access_history_limit: number | null;
  access_history_offset: number;
  verification: ChainOfCustodyVerification;
}
